File: sdks/jvm/golem-java/src/main/java/cloud/golem/java/types/Either.java
/*
 * Copyright 2024-2025 Golem Cloud
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cloud.golem.java.types;

import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.function.Function;

/**
 * Represents a value that can be one of two types: Left or Right.
 *
 * <p>By convention, Left is used for "failure" or "alternative" cases,
 * while Right is the "success" or "primary" case.
 *
 * @param <L> The left type
 * @param <R> The right type
 */
public final class Either<L, R> {

    private final Object value;
    private final boolean isRight;

    private Either(Object value, boolean isRight) {
        this.value = value;
        this.isRight = isRight;
    }

    /**
     * Create a Left value.
     *
     * @param value The left value
     * @param <L> The left type
     * @param <R> The right type
     * @return An Either containing the left value
     */
    public static <L, R> Either<L, R> left(L value) {
        return new Either<>(value, false);
    }

    /**
     * Create a Right value.
     *
     * @param value The right value
     * @param <L> The left type
     * @param <R> The right type
     * @return An Either containing the right value
     */
    public static <L, R> Either<L, R> right(R value) {
        return new Either<>(value, true);
    }

    /**
     * Check if this is a Left value.
     *
     * @return true if Left
     */
    public boolean isLeft() {
        return !isRight;
    }

    /**
     * Check if this is a Right value.
     *
     * @return true if Right
     */
    public boolean isRight() {
        return isRight;
    }

    /**
     * Get the Left value.
     *
     * @return The Left value
     * @throws NoSuchElementException If this is a Right value
     */
    @SuppressWarnings("unchecked")
    public L unwrapLeft() {
        if (isRight) {
            throw new NoSuchElementException("Called unwrapLeft on a Right: " + value);
        }
        return (L) value;
    }

    /**
     * Get the Right value.
     *
     * @return The Right value
     * @throws NoSuchElementException If this is a Left value
     */
    @SuppressWarnings("unchecked")
    public R unwrapRight() {
        if (!isRight) {
            throw new NoSuchElementException("Called unwrapRight on a Left: " + value);
        }
        return (R) value;
    }

    /**
     * Transform the Right value if present.
     *
     * @param mapper The transformation function
     * @param <A> The result type
     * @return An Either with the transformed Right value
     */
    @SuppressWarnings("unchecked")
    public <A> Either<L, A> map(Function<? super R, ? extends A> mapper) {
        if (isRight) {
            return right(mapper.apply((R) value));
        }
        return left((L) value);
    }

    /**
     * Transform the Left value if present.
     *
     * @param mapper The transformation function
     * @param <A> The result type
     * @return An Either with the transformed Left value
     */
    @SuppressWarnings("unchecked")
    public <A> Either<A, R> mapLeft(Function<? super L, ? extends A> mapper) {
        if (!isRight) {
            return left(mapper.apply((L) value));
        }
        return right((R) value);
    }

    /**
     * Transform the Right value with a function that returns an Either.
     *
     * @param mapper The transformation function
     * @param <A> The result type
     * @return The result of the transformation
     */
    @SuppressWarnings("unchecked")
    public <A> Either<L, A> flatMap(Function<? super R, Either<L, A>> mapper) {
        if (isRight) {
            return mapper.apply((R) value);
        }
        return left((L) value);
    }

    /**
     * Handle both cases and produce a single result.
     *
     * @param leftMapper Function to handle Left case
     * @param rightMapper Function to handle Right case
     * @param <A> The result type
     * @return The result of the matching function
     */
    @SuppressWarnings("unchecked")
    public <A> A fold(
            Function<? super L, ? extends A> leftMapper,
            Function<? super R, ? extends A> rightMapper) {
        if (isRight) {
            return rightMapper.apply((R) value);
        }
        return leftMapper.apply((L) value);
    }

    /**
     * Swap Left and Right.
     *
     * @return An Either with Left and Right swapped
     */
    @SuppressWarnings("unchecked")
    public Either<R, L> swap() {
        if (isRight) {
            return left((R) value);
        }
        return right((L) value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Either<?, ?> either = (Either<?, ?>) o;
        if (isRight != either.isRight) return false;
        return Objects.equals(value, either.value);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value, isRight);
    }

    @Override
    public String toString() {
        return isRight ? "Either.right(" + value + ")" : "Either.left(" + value + ")";
    }
}

File: sdks/python/golem_sdk/types/either.py
# Copyright 2024-2025 Golem Cloud
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Either type for representing one of two values."""

from typing import Callable, Generic, TypeVar, Union

L = TypeVar("L")
R = TypeVar("R")
A = TypeVar("A")
B = TypeVar("B")


class Either(Generic[L, R]):
    """
    Represents a value that can be one of two types: Left or Right.

    This is similar to Haskell's Either type or Scala's Either type.
    By convention, Left is used for "failure" or "alternative" cases,
    while Right is the "success" or "primary" case.

    Example:
        ```python
        left_value: Either[str, int] = Either.left("error")
        right_value: Either[str, int] = Either.right(42)

        # Check which side
        if right_value.is_right():
            print(right_value.unwrap_right())  # 42

        # Map over the right value
        doubled = right_value.map(lambda x: x * 2)  # Either.right(84)

        # Handle both cases
        result = right_value.fold(
            left=lambda s: f"Error: {s}",
            right=lambda n: f"Got: {n}",
        )
        ```
    """

    __slots__ = ("_value", "_is_right")

    def __init__(self, value: Union[L, R], is_right: bool) -> None:
        self._value = value
        self._is_right = is_right

    @classmethod
    def left(cls, value: L) -> "Either[L, R]":
        """Create a Left value."""
        return cls(value, False)

    @classmethod
    def right(cls, value: R) -> "Either[L, R]":
        """Create a Right value."""
        return cls(value, True)

    def is_left(self) -> bool:
        """Check if this is a Left value."""
        return not self._is_right

    def is_right(self) -> bool:
        """Check if this is a Right value."""
        return self._is_right

    def unwrap_left(self) -> L:
        """
        Get the Left value.

        Raises:
            ValueError: If this is a Right value.
        """
        if self._is_right:
            raise ValueError(f"Called unwrap_left on a Right: {self._value}")
        return self._value  # type: ignore

    def unwrap_right(self) -> R:
        """
        Get the Right value.

        Raises:
            ValueError: If this is a Left value.
        """
        if not self._is_right:
            raise ValueError(f"Called unwrap_right on a Left: {self._value}")
        return self._value  # type: ignore

    def map(self, f: Callable[[R], A]) -> "Either[L, A]":
        """Transform the Right value if present."""
        if self._is_right:
            return Either.right(f(self._value))  # type: ignore
        return Either.left(self._value)  # type: ignore

    def map_left(self, f: Callable[[L], A]) -> "Either[A, R]":
        """Transform the Left value if present."""
        if not self._is_right:
            return Either.left(f(self._value))  # type: ignore
        return Either.right(self._value)  # type: ignore

    def flat_map(self, f: Callable[[R], "Either[L, A]"]) -> "Either[L, A]":
        """Transform the Right value with a function that returns an Either."""
        if self._is_right:
            return f(self._value)  # type: ignore
        return Either.left(self._value)  # type: ignore

    def fold(
        self,
        left: Callable[[L], A],
        right: Callable[[R], A],
    ) -> A:
        """Handle both cases and produce a single result."""
        if self._is_right:
            return right(self._value)  # type: ignore
        return left(self._value)  # type: ignore

    def swap(self) -> "Either[R, L]":
        """Swap Left and Right."""
        if self._is_right:
            return Either.left(self._value)  # type: ignore
        return Either.right(self._value)  # type: ignore

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Either):
            return NotImplemented
        if self._is_right != other._is_right:
            return False
        return self._value == other._value

    def __repr__(self) -> str:
        if self._is_right:
            return f"Either.right({self._value!r})"
        return f"Either.left({self._value!r})"

    def __hash__(self) -> int:
        if self._is_right:
            return hash(("right", self._value))
        return hash(("left", self._value))


# Convenience functions
def left(value: L) -> Either[L, R]:
    """Create a Left value."""
    return Either.left(value)


def right(value: R) -> Either[L, R]:
    """Create a Right value."""
    return Either.right(value)
